File: eval-notebooks/evaluator.py
from dotenv import load_dotenv
from openai import AsyncOpenAI
import os
import asyncio
import json


class Evaluator:
    def __init__(self, *model_names):

        load_dotenv()

        OPENROUTER_API_KEY = os.getenv("OPENROUTER_API_KEY")
        if not OPENROUTER_API_KEY:
            raise ValueError("OPENROUTER_API_KEY not found in .env file or environment variables.")
        
        self.client = AsyncOpenAI(
            base_url="https://openrouter.ai/api/v1",
            api_key=OPENROUTER_API_KEY
        )

        self.candidate_models = [
            # PAID
            {"name": "Anthropic: Claude Sonnet 4.5", "openrouter": "anthropic/claude-sonnet-4.5"},
            {"name": "OpenAI: GPT-4o", "openrouter": "openai/gpt-4o"},
            {"name": "OpenAI: GPT-4o Mini", "openrouter": "openai/gpt-4o-mini"},
            # FREE
            {"name": "TNG: DeepSeek R1T2 Chimera (free)", "openrouter": "tngtech/deepseek-r1t2-chimera:free"},
            {"name": "Meta: Llama 3.3 70B Instruct (free)", "openrouter": "meta-llama/llama-3.3-70b-instruct:free"},
            {"name": "OpenAI: gpt-oss-20b (free)", "openrouter": "openai/gpt-oss-20b:free"},
            {"name": "AllenAI: Molmo2 8B (free)", "openrouter": "allenai/molmo-2-8b:free"},
            {"name": "StepFun: Step 3.5 Flash (free)", "openrouter": "stepfun/step-3.5-flash:free"}
        ]

        self.model_names = model_names



    async def query_model(self, model_name: str, query: str, role="user"):
        '''
        Queries a single model using the models in 'candidate_models'

        :param model_name: The name of the model (from candidate_models "name" field)
        '''
        model_dict = next((m for m in self.candidate_models if m["name"] == model_name), None)
        if model_dict is None:
            return {"model": model_name, "response": f"Error: Model '{model_name}' not found in candidate_models"}
        
        try:
            response = await self.client.chat.completions.create(
                model=model_dict["openrouter"],
                messages=[{"role" : role, "content" : query}],
                temperature=1
            )
            content = response.choices[0].message.content
            return {"model": model_name, "response": content}
        except Exception as e:
            return {"model": model_name, "response": str(e)}
    
    async def query_models(self, model_names: list[str], queries: list[str], role="user"):
        '''
        Queries multiple models asynchronously

        :param model_names: List of model names (from candidate_models "name" field)
        '''
        coroutines = [self.query_model(model_names[i], queries[i], role=role) for i in range(len(model_names))]
        results = await asyncio.gather(*coroutines)
        return results



    def extract_outermost_json(self, text: str):
        '''
        Extracts the first outermost JSON value (object or array) from arbitrary text.
        
        :param text: The text to extract JSON from
        '''
        if not text:
            return None

        # Strip markdown code fences (``` or ```json)
        s = text.strip()
        if s.startswith("```"):
            lines = s.splitlines()
            s = "\n".join(line for line in lines if not line.strip().startswith("```")).strip()

        start = None
        stack = []  # holds '{' or '['
        in_string = False
        escape = False

        pairs = {"{": "}", "[": "]"}

        for i, ch in enumerate(s):
            if escape:
                escape = False
                continue

            if ch == "\\":
                escape = True
                continue

            if ch == '"':
                in_string = not in_string
                continue

            if in_string:
                continue

            if ch in "{[":
                if not stack:
                    start = i
                stack.append(ch)

            elif ch in "}]":
                if not stack:
                    continue
                opener = stack[-1]
                if pairs[opener] == ch:
                    stack.pop()
                    if not stack and start is not None:
                        candidate = s[start:i + 1]
                        try:
                            return json.loads(candidate)
                        except json.JSONDecodeError:
                            # if this block isn't valid JSON, keep scanning
                            start = None
                            continue
                else:
                    # mismatched bracket; reset search
                    stack.clear()
                    start = None

        return None

File: lm-compass/lib/evaluation/evaluator.ts
/**
 * Base Evaluator class - mirrors evaluator.py
 * Provides core functionality for querying models and extracting JSON from responses
 */

import type { OpenAI } from "openai";

/**
 * Model configuration for OpenRouter
 */
interface CandidateModel {
  name: string;
  openrouter: string;
}

/**
 * Query response containing model name and response content
 */
export interface ModelQueryResponse {
  model: string;
  response: string;
}

/**
 * Base Evaluator class
 */
export class Evaluator {
  protected client: OpenAI;
  protected modelNames: string[];
  protected candidateModels: CandidateModel[];

  constructor(client: OpenAI, ...modelNames: string[]) {
    this.client = client;
    this.modelNames = modelNames;

    // Candidate models configuration - mirrors evaluator.py
    this.candidateModels = [
      // PAID
      {
        name: "Anthropic: Claude Sonnet 4.5",
        openrouter: "anthropic/claude-sonnet-4.5",
      },
      { name: "OpenAI: GPT-4o", openrouter: "openai/gpt-4o" },
      { name: "OpenAI: GPT-4o Mini", openrouter: "openai/gpt-4o-mini" },
      // FREE
      {
        name: "StepFun: Step 3.5 Flash (free)",
        openrouter: "stepfun/step-3.5-flash:free",
      },
      {
        name: "TNG: DeepSeek R1T2 Chimera (free)",
        openrouter: "tngtech/deepseek-r1t2-chimera:free",
      },
      {
        name: "LiquidAI: LFM2.5-1.2B-Thinking (free)",
        openrouter: "liquid/lfm-2.5-1.2b-thinking:free",
      },
    ];
  }

  /**
   * Query a single model
   * @param modelName - The name of the model (from candidateModels "name" field)
   * @param query - The query to send to the model
   * @param role - The role of the message (default: "user")
   * @returns Object with model name and response content
   */
  async queryModel(
    modelName: string,
    query: string,
    role: "system" | "user" | "assistant" = "user",
  ): Promise<ModelQueryResponse> {
    const modelDict = this.candidateModels.find((m) => m.name === modelName);
    if (!modelDict) {
      return {
        model: modelName,
        response: `Error: Model '${modelName}' not found in candidate_models`,
      };
    }

    try {
      const response = await this.client.chat.completions.create({
        model: modelDict.openrouter,
        messages: [{ role, content: query }],
        temperature: 1,
      });
      const content = response.choices[0].message.content;
      return {
        model: modelName,
        response: content || "",
      };
    } catch (e) {
      return {
        model: modelName,
        response: String(e),
      };
    }
  }

  /**
   * Query multiple models asynchronously
   * @param modelNames - List of model names (from candidateModels "name" field)
   * @param queries - List of queries (one per model)
   * @param role - The role of the message (default: "user")
   * @returns Array of query responses
   */
  async queryModels(
    modelNames: string[],
    queries: string[],
    role: "system" | "user" | "assistant" = "user",
  ): Promise<ModelQueryResponse[]> {
    const promises = modelNames.map((modelName, i) =>
      this.queryModel(modelName, queries[i], role),
    );
    return Promise.all(promises);
  }

  /**
   * Extracts the first outermost JSON value (object or array) from arbitrary text
   * Mirrors the Python implementation
   * @param text - The text to extract JSON from
   * @returns Parsed JSON object/array, or null if not found
   */
  extractOutermostJson(text: string): unknown {
    if (!text) {
      return null;
    }

    // Strip markdown code fences (``` or ```json)
    let s = text.trim();
    if (s.startsWith("```")) {
      const lines = s.split("\n");
      s = lines
        .filter((line) => !line.trim().startsWith("```"))
        .join("\n")
        .trim();
    }

    let start: number | null = null;
    const stack: string[] = []; // holds '{' or '['
    let inString = false;
    let escape = false;

    const pairs: Record<string, string> = { "{": "}", "[": "]" };

    for (let i = 0; i < s.length; i++) {
      const ch = s[i];

      if (escape) {
        escape = false;
        continue;
      }

      if (ch === "\\") {
        escape = true;
        continue;
      }

      if (ch === '"') {
        inString = !inString;
        continue;
      }

      if (inString) {
        continue;
      }

      if (ch === "{" || ch === "[") {
        if (stack.length === 0) {
          start = i;
        }
        stack.push(ch);
      } else if (ch === "}" || ch === "]") {
        if (stack.length === 0) {
          continue;
        }
        const opener = stack[stack.length - 1];
        if (pairs[opener] === ch) {
          stack.pop();
          if (stack.length === 0 && start !== null) {
            const candidate = s.substring(start, i + 1);
            try {
              return JSON.parse(candidate);
            } catch {
              // JSON.parse may fail if the string contains invalid escape
              // sequences (e.g. LaTeX \( \) \[ \] produced by LLMs).
              // Replace any backslash NOT followed by a recognised JSON
              // escape character with a double-backslash and retry.
              try {
                const sanitized = candidate.replace(
                  /\\(?!["\\\/bfnrt]|u[0-9a-fA-F]{4})/g,
                  "\\\\",
                );
                return JSON.parse(sanitized);
              } catch {
                // Still not valid JSON — keep scanning for the next candidate
                start = null;
                continue;
              }
            }
          }
        } else {
          // mismatched bracket; reset search
          stack.length = 0;
          start = null;
        }
      }
    }

    return null;
  }
}
